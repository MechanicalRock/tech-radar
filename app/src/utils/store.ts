import * as reducers from '../reducers';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';

const store = createStore(
  combineReducers<ReduxState>(reducers),
  applyMiddleware(thunk)
);

export default store;
