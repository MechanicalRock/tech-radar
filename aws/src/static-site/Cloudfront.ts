import * as s3 from '@aws-cdk/aws-s3';
import { Stack } from '@aws-cdk/cdk';
import {
  CloudFrontWebDistribution,
  CloudFrontWebDistributionProps,
  CfnCloudFrontOriginAccessIdentity,
  PriceClass,
  CloudFrontAllowedMethods,
  CloudFrontAllowedCachedMethods
} from '@aws-cdk/aws-cloudfront';
import { Bucket } from '@aws-cdk/aws-s3';
import { Output } from '@aws-cdk/cdk';

export interface CloudfrontProps {
  stack: Stack;
  logicalName: string;
  siteBucket: s3.Bucket;
  originAccessIdentity: CfnCloudFrontOriginAccessIdentity;
}

export class Cloudfront {
  distribution: CloudFrontWebDistribution;

  constructor(private props: CloudfrontProps) {
    this.distribution = this.create();
    this.createOutputs();
  }

  private create(): CloudFrontWebDistribution {
    const distributionProps: CloudFrontWebDistributionProps = {
      comment: 'Cloud front distribution for static website',
      aliasConfiguration: {
        acmCertRef:
          'arn:aws:acm:us-east-1:598112752826:certificate/da98edf4-7dc7-4ee8-95d6-82d96bc1a13b',
        names: ['techradar.cloud', 'www.techradar.cloud']
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.props.siteBucket,
            originAccessIdentity: this.props.originAccessIdentity
          },
          behaviors: [
            {
              isDefaultBehavior: true,
              compress: true,
              allowedMethods: CloudFrontAllowedMethods.GET_HEAD_OPTIONS,
              cachedMethods: CloudFrontAllowedCachedMethods.GET_HEAD_OPTIONS,
              minTtlSeconds: 3600
            }
          ]
        }
      ],
      errorConfigurations: [
        {
          errorCode: 404,
          responseCode: 200,
          responsePagePath: '/'
        }
      ],
      loggingConfig: {
        bucket: this.createLoggingBucket()
      },
      priceClass: PriceClass.PriceClassAll
    };
    return new CloudFrontWebDistribution(
      this.props.stack,
      this.props.logicalName,
      distributionProps
    );
  }

  private createOutputs() {
    new Output(this.props.stack, 'DistributionDomainName', {
      value: this.distribution.domainName,
      disableExport: true
    });
  }

  private createLoggingBucket(): Bucket {
    return new Bucket(this.props.stack, 'LogsBucket', {
      encryption: s3.BucketEncryption.S3Managed
    });
  }
}
