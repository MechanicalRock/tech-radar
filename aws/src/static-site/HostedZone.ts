import { Stack, Output } from '@aws-cdk/cdk';
import {
  PublicHostedZone,
  AliasRecord,
  IAliasRecordTarget
} from '@aws-cdk/aws-route53';

export interface HostedZoneProps {
  stack: Stack;
  fullyQualifiedDomainName: string;
}

export class HostedZone {
  hostedZone: PublicHostedZone;
  alias: IAliasRecordTarget;

  constructor(private props: HostedZoneProps, alias: IAliasRecordTarget) {
    this.alias = alias;
    this.create();
    this.publishOutputs();
  }

  create(): void {
    this.hostedZone = new PublicHostedZone(this.props.stack, 'HostedZone', {
      zoneName: this.props.fullyQualifiedDomainName,
      comment: `Hosted zone for domain: ${this.props.fullyQualifiedDomainName}`
    });
    new AliasRecord(this.props.stack, 'CloudFrontAlias', {
      zone: this.hostedZone,
      recordName: this.props.fullyQualifiedDomainName,
      target: this.alias
    });
  }

  publishOutputs(): void {
    new Output(this.props.stack, 'HostedZoneId', {
      value: this.hostedZone.hostedZoneId
    });
  }
}
