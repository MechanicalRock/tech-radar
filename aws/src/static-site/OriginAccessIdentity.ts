import { Stack } from '@aws-cdk/cdk';
import { CfnCloudFrontOriginAccessIdentity } from '@aws-cdk/aws-cloudfront';

export class OriginAccessIdentity {
  private originAccessIdentity: CfnCloudFrontOriginAccessIdentity;
  constructor(private stack: Stack) {
    this.originAccessIdentity = this.create();
  }

  getResource(): CfnCloudFrontOriginAccessIdentity {
    return this.originAccessIdentity;
  }

  private create(): CfnCloudFrontOriginAccessIdentity {
    return new CfnCloudFrontOriginAccessIdentity(
      this.stack,
      'OriginAccessIdentity',
      {
        cloudFrontOriginAccessIdentityConfig: {
          comment:
            'Cloud front user which will be allowed to access the site s3 bucket'
        }
      }
    );
  }
}
