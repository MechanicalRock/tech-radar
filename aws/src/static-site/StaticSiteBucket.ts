import { Stack, Parameter } from '@aws-cdk/cdk';
import * as s3 from '@aws-cdk/aws-s3';
import { PolicyStatement, CanonicalUserPrincipal } from '@aws-cdk/aws-iam';

export interface StaticSiteBucketProps {
  stack: Stack;
  logicalName: string;
  originAccessIdentityCanonicalId: string;
}

export class StaticSiteBucket {
  bucketParam: Parameter;
  bucket: s3.Bucket;

  constructor(private props: StaticSiteBucketProps) {
    this.bucket = this.create();
  }

  private create() {
    const bucket = new s3.Bucket(this.props.stack, this.props.logicalName, {
      encryption: s3.BucketEncryption.S3Managed
    });

    const statement = new PolicyStatement();
    statement.addPrincipal(
      new CanonicalUserPrincipal(this.props.originAccessIdentityCanonicalId)
    );
    statement.addAction('s3:GetObject');
    statement.addResource(bucket.bucketArn + '/*');
    bucket.addToResourcePolicy(statement);
    return bucket;
  }
}
