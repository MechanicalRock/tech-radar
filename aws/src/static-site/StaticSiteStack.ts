import * as cdk from '@aws-cdk/cdk';
import { Parameter } from '@aws-cdk/cdk';
import { StaticSiteBucket } from './StaticSiteBucket';
import { OriginAccessIdentity } from './OriginAccessIdentity';
import { Cloudfront } from './Cloudfront';
import { HostedZoneAlias } from './HostedZone';

export class StaticSiteStack extends cdk.Stack {
  bucketParameter: Parameter;

  constructor(parent: cdk.App, stackName: string, props?: cdk.StackProps) {
    super(parent, stackName, props);
    this.create();
  }

  create(): void {
    const originAccessIdentityResource = new OriginAccessIdentity(
      this
    ).getResource();

    const siteBucket = new StaticSiteBucket({
      stack: this,
      logicalName: 'StaticSiteBucket',
      originAccessIdentityCanonicalId:
        originAccessIdentityResource.cloudFrontOriginAccessIdentityS3CanonicalUserId
    });

    const { distribution } = new Cloudfront({
      stack: this,
      siteBucket: siteBucket.bucket,
      logicalName: 'static-site-cloudfront',
      originAccessIdentity: originAccessIdentityResource
    });

    new HostedZoneAlias(
      {
        stack: this,
        fullyQualifiedDomainName: 'techradar.cloud'
      },
      distribution
    );
  }
}
