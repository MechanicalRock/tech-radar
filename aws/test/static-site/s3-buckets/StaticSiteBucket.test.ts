import { Stack } from '@aws-cdk/cdk';
import { StaticSiteBucket } from '../../../src/static-site/StaticSiteBucket';
import * as yaml from '../../utils/Yaml';
import { OriginAccessIdentity } from '../../../src/static-site/OriginAccessIdentity';

describe('StaticSiteBucket', () => {
    it('should match expected template snapshot', () => {
        const stack = new Stack();
        const originAccessIdentity = new OriginAccessIdentity(stack);

        // When
        new StaticSiteBucket({
            stack,
            logicalName: 'MyStaticSiteBucket',
            originAccessIdentityCanonicalId: originAccessIdentity.getResource().cloudFrontOriginAccessIdentityS3CanonicalUserId
        });

        // Then
        expect(yaml.toYamlString(stack.toCloudFormation())).toMatchSnapshot();
    })
})