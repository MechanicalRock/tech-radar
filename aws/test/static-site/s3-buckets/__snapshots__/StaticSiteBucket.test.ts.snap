// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticSiteBucket should match expected template snapshot 1`] = `
"Resources:
  OriginAccessIdentity:
    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity'
    Properties:
      CloudFrontOriginAccessIdentityConfig:
        Comment: Cloud front user which will be allowed to access the site s3 bucket
  MyStaticSiteBucket0830CFE7:
    Type: 'AWS::S3::Bucket'
    Properties:
      BucketEncryption:
        ServerSideEncryptionConfiguration:
          - ServerSideEncryptionByDefault:
              SSEAlgorithm: AES256
    DeletionPolicy: Retain
  MyStaticSiteBucketPolicy88B97D28:
    Type: 'AWS::S3::BucketPolicy'
    Properties:
      Bucket:
        Ref: MyStaticSiteBucket0830CFE7
      PolicyDocument:
        Statement:
          - Action: 's3:GetObject'
            Effect: Allow
            Principal:
              CanonicalUser:
                'Fn::GetAtt':
                  - OriginAccessIdentity
                  - S3CanonicalUserId
            Resource:
              'Fn::Join':
                - ''
                - - 'Fn::GetAtt':
                      - MyStaticSiteBucket0830CFE7
                      - Arn
                  - /*
        Version: '2012-10-17'
"
`;
